adcommon/yldap.py

```diff
@@ -175,3 +180,126 @@ def ldap_delete(self, *args):
+
+        self.schema['objectClasses'] = {}
+        for objectClass in results['objectClasses']:
+            m = re.match(b'\(\s+(?P<id>[0-9\.]+)\s+NAME\s+\'(?P<name>[\-\w]+)\'\s+(SUP\s+(?P<superior>[\-\w]+)\s+)?(?P<type>\w+)\s+(MUST\s+\((?P<must>[^\)]*)\)\s+)?(MAY\s+\((?P<may>[^\)]*)\)\s+)?\)', objectClass)
+            if m:
+                name = m.group('name')
+                self.schema['objectClasses'][name] = {}
+                self.schema['objectClasses'][name]['id'] = m.group('id')
+                self.schema['objectClasses'][name]['superior'] = m.group('superior')
+                self.schema['objectClasses'][name]['inferior'] = self.__find_inferior_classes(name.decode())
+                self.schema['objectClasses'][name]['type'] = m.group('type')
+                self.schema['objectClasses'][name]['must'] = m.group('must').strip().split(b' $ ') if m.group('must') else []
+                self.schema['objectClasses'][name]['may'] = m.group('may').strip().split(b' $ ') if m.group('may') else []
+            else:
+                raise ldap.LDAPError('Failed to parse objectClass: %s' % objectClass.decode())
+
+        self.schema['dITContentRules'] = {}
+        for dITContentRule in results['dITContentRules']:
+            m = re.match(b'\(\s+(?P<id>[0-9\.]+)\s+NAME\s+\'(?P<name>[\-\w]+)\'\s*(AUX\s+\((?P<aux>[^\)]*)\))?\s*(MUST\s+\((?P<must>[^\)]*)\)\s+)?\s*(MAY\s+\((?P<may>[^\)]*)\))?\s*(NOT\s+\((?P<not>[^\)]*)\))?\s*\)', dITContentRule)
+            if m:
+                name = m.group('name')
+                self.schema['dITContentRules'][name] = {}
+                self.schema['dITContentRules'][name]['id'] = m.group('id')
+                self.schema['dITContentRules'][name]['must'] = m.group('must').strip().split(b' $ ') if m.group('must') else []
+                self.schema['dITContentRules'][name]['may'] = m.group('may').strip().split(b' $ ') if m.group('may') else []
+                self.schema['dITContentRules'][name]['aux'] = m.group('aux').strip().split(b' $ ') if m.group('aux') else []
+                self.schema['dITContentRules'][name]['not'] = m.group('not').strip().split(b' $ ') if m.group('not') else []
+            else:
+                raise ldap.LDAPError('Failed to parse dITContentRule: %s' % dITContentRule.decode())
+
+    def __constructed_attributes(self):
+        # ADSI Hides constructed attributes, since they can't be modified.
+        search = '(&(systemFlags:1.2.840.113556.1.4.803:=4)(ObjectClass=attributeSchema))'
+        container = 'CN=Schema,CN=Configuration,%s' % self.realm_dn
+        ret = self.ldap_search(container, SCOPE_ONELEVEL, search, ['lDAPDisplayName'])
+        return [a[-1]['lDAPDisplayName'][-1] for a in ret]
+
+    def __timestamp(self, val):
+        return str(datetime.strptime(val.decode(), '%Y%m%d%H%M%S.%fZ'))
+
+    def __display_value_each(self, syntax, key, val):
+        if syntax == b'1.3.6.1.4.1.1466.115.121.1.24':
```

Comment:
this is very magic string. I think it really deserve some comment explaining it and also ideally link to some documentation.

Reply:
This black magic is explained in rfc4517 :)
Sorry for the confusion, I'll add this in comments.